src/box/txn_limbo.c

```diff
@@ -1236,10 +1316,8 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
 	case IPROTO_RAFT_PROMOTE:
 	case IPROTO_RAFT_DEMOTE: {
 		assert(limbo->svp_confirmed_lsn == -1);
-		limbo->svp_confirmed_lsn = limbo->confirmed_lsn;
-		limbo->confirmed_lsn = req->lsn;
-		vclock_reset(&limbo->confirmed_vclock, limbo->owner_id,
-			     req->lsn);
```

Comment:
Why do we no longer reset `confirmed_vclock` here and below?

Reply:
Because:
1) It is always updated in `txn_limbo_read_confirm` as part of `txn_limbo_read_promote` during `txn_limbo_req_commit`. If I leave the reset here, I will end up updating it to the same value, which will trigger an assertion. 
2) AFAIC, an LSN coming from a PROMOTE/DEMOTE request must always be `vclock_follow`ed rather than `vclock_reset`.

Reply:
> AFAIC, an LSN coming from a PROMOTE/DEMOTE request must always be vclock_followed rather than vclock_reset, since it can only monotonically grow

It grows and shrinks the same as `confirmed_vclock`. If it would only grow, then you would be able to leave the assertion that the confirmed vclock always only grows. But evidently it doesn't. No?

Besides, from what I see, PROMOTE only has `confirmed_vclock` set in `synchro_request` on checkpoint. On non-checkpoint PROMOTE this field is NULL. Which means that being sent to a remote host it won't have this vclock set, and `txn_limbo_req_commit()` won't do `vclock_copy()`. Would you please investigate why does it pass the tests, when you deleted `vclock_reset()`? That doesn't look right. Either this isn't covered by any tests, or vclock-copy happens somewhere somehow, or you should be able to leave `confirmed_vclock` also monotonically growing.

Reply:
I forgot to leave an update to my comment: I discussed this with @sergepetrenko and we came to a conclusion that `vclock_reset` here is needed solely for the case when `req->lsn` is the same as `confirmed_lsn` (`vclock_follow` prohibits passing a value equal to the current one).

So, these two snippets would actually be equivalent in `txn_limbo_req_prepare`:
```c
limbo->confirmed_lsn = req->lsn;
vclock_reset(&limbo->confirmed_vclock, limbo->owner_id, req->lsn);
```

```c
if (limbo->confirmed_lsn < req_lsn) {
    limbo->confirmed_lsn = req->lsn;
    vclock_follow(&limbo->confirmed_vclock, limbo->owner_id, req->lsn);
}
```

The second snippet is exactly what `txn_limbo_read_confirm` does:
https://github.com/tarantool/tarantool/blob/5325abd3441ecb4b589799c32ec181d88724b8a8/src/box/txn_limbo.c#L551-L554

So all I did is delayed the `confirmed_lsn` and `confirmed_vclock` updates to `txn_limbo_read_confirm`.

> It grows and shrinks the same as confirmed_vclock. If it would only grow, then you would be able to leave the assertion that the confirmed vclock always only grows. But evidently it doesn't. No?

It never shrinks. It monotonically grows — not _strictly_ monotonically though. The problem is with the lack of strictness, since a promote request with `req->lsn` equal to the `confirmed_lsn` and equal to the `confirmed_vclock` component could arrive.

> Besides, from what I see, PROMOTE only has confirmed_vclock set in synchro_request on checkpoint.

That's not correct, it is updated in `txn_limbo_read_confirm`, as I have pointed out above.

> Would you please investigate why does it pass the tests, when you deleted `vclock_reset()`?

I hope my explanation above is convincing enough. At least I managed to convince @sergepetrenko.

Reply:
Aaah, so previously `confirmed_lsn` was already "volatile". This is why it could legally go back, and the vclock too. But now it is always updated only on disk. So it is either monotonic, or we get a split brain. Beautiful!

> That's not correct, it is updated in txn_limbo_read_confirm, as I have pointed out above.

I didn't speak about update. I was wondering about `synchro_request`. In this object it is only set on checkpoint `PROMOTE`. Which means `confirmed_vclock` is never sent around besides on join probably. But it is irrelevant now. I understand. Thanks for the explanation.

Reply:
Btw, does this all mean `svp_confirmed_lsn` is not needed anymore? For example, look at this:
```C
	case IPROTO_RAFT_PROMOTE:
	case IPROTO_RAFT_DEMOTE: {
		assert(limbo->svp_confirmed_lsn == -1);
		limbo->svp_confirmed_lsn = limbo->volatile_confirmed_lsn;
		limbo->volatile_confirmed_lsn = req->lsn;
		txn_limbo_update_system_spaces_is_sync_state(
			limbo, req, /*is_rollback=*/false);
		break;
	}
```
We could as well completely drop `svp_confirmed_lsn`, and on promotion failure we could restore `limbo->volatile_confirmed_lsn = limbo->confirmed_lsn`.

Reply:
> Btw, does this all mean svp_confirmed_lsn is not needed anymore? For example, look at this

No, imagine the following scenario the current limbo owner:
```
Initially: volatile_confirmed_lsn = 1, confirmed_lsn = 1
TX commit: -> volatile_confirmed = 2, confirmed_lsn = 1
Limbo fiber: doesn't get woken up for a while
Promote request from a new leader: volatile_confirmed_lsn = 1, confirmed_lsn = 1
Promote request fails: volatile_confirmed_lsn = confirmed_lsn = 1, confirmed_lsn = 1
```

But after the promote request fails `volatile_confirmed_lsn` is supposed to return to the value 2.